dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/QuayImageRegistry.java

```diff
@@ -65,45 +69,72 @@
     private static final Logger LOG = LoggerFactory.getLogger(QuayImageRegistry.class);
 
     private final Token quayToken;
-    private final ApiClient apiClient;
     private final BuildApi buildApi;
     private final RepositoryApi repositoryApi;
     private final UserApi userApi;
+    private final TagApi tagApi;
 
     public QuayImageRegistry(final Token quayToken) {
         this.quayToken = quayToken;
-
-        apiClient = Configuration.getDefaultApiClient();
+        ApiClient apiClient = Configuration.getDefaultApiClient();
         apiClient.addDefaultHeader("Authorization", "Bearer " + quayToken.getContent());
         this.buildApi = new BuildApi(apiClient);
         this.repositoryApi = new RepositoryApi(apiClient);
         this.userApi = new UserApi(apiClient);
+        this.tagApi = new TagApi((apiClient));
+
     }
 
+    private List<QuayTag> getAllQuayTags(String repository) throws ApiException {
+        List<QuayTag> allQuayTags = new ArrayList<>();
+        // Completely arbitrary maxPageSize in the weird event that Quay.io's pagination results in an infinite loop or something
+        final int maxPageSize = 100;
+        for (int page = 1; page < Integer.MAX_VALUE; page++) {
+            InlineResponse2002 inlineResponse2002 = tagApi.listRepoTags(repository, page, maxPageSize, null, true);
+            List<QuayTag> quayTags = inlineResponse2002.getTags();
+            allQuayTags.addAll(quayTags);
+            if (!inlineResponse2002.isHasAdditional()) {
+                break;
+            }
+        }
+        return allQuayTags;
+    }
+
+
     @Override
     public List<Tag> getTags(Tool tool) {
         LOG.info(quayToken.getUsername() + " ======================= Getting tags for: {}================================", tool.getPath());
-
+        final String repo = tool.getNamespace() + '/' + tool.getName();
         final List<Tag> tags = new ArrayList<>();
         final Optional<QuayRepo> toolFromQuay = getToolFromQuay(tool);
         if (toolFromQuay.isPresent()) {
             final QuayRepo quayRepo = toolFromQuay.get();
             final Map<String, QuayTag> tagsFromRepo = quayRepo.getTags();
-            for (QuayTag tagItem : tagsFromRepo.values()) {
+            final int maxQuayTagsReturnedByRepo = 500;
+            if (tagsFromRepo.size() == maxQuayTagsReturnedByRepo) {
```

Comment:
Seems possible to simplify. 
If I understand correctly, this loop is handling the case when we need to paginate. The loop below is when you do not. 
But all `getAllQuayTags` is doing is putting all tags into one big collection. 
It should be possible to simplify by modifying `getAllQuayTags` to be tolerant of results that are one page.

Reply:
The loop below is for converting all QuayTags to Tags. getAllQuayTags works even if there's one page (not that it'll ever get to that point).

Reply:
Not sure if we're talking about the same thing. 
To me it looks like the loop on line 117 and the look on 129 are the same, they just run across different collections. It should be possible to 
1) Have one method with a conditional in it to build the right collection in the two cases 
2) run one loop on the built collection

Reply:
Changed a few things, but it's one loop now